# Exercise 4.5 TEXT

In [1]:
# Importing libraries
import numpy as np
import pandas as pd
import os

In [5]:
# Dataframes
path = r'C:\Users\nmiqu\Desktop\DATA\CareerFoundry\Data Analytics Immersion\April-2024 Instacart Baket Analysis\02 Data'
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
# Create a dataframe
df_test = pd.DataFrame()

In [7]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
# Missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
df_prods.shape

(49693, 5)

In [18]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
df_prods_clean.shape

(49677, 5)

In [21]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
df_prods_clean.shape

(49677, 5)

In [26]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [27]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [28]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Prepared Data', 'products_clean.csv'))

# Exercise 4.5 TASK

In [30]:
df_prods_clean = df_prods_clean_no_dups

In [34]:
# Step 2
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  hour_order_is_made  days_since_prior_order  
count        3.421083e+06        3.421083e+06            3.214874e+06  
mean         2.776219e+00        1.345202e+01            1.111484e+01  
std          2.046829e+00        4.226088e+00            9.206737e+00  
min          0.000000e+00        0.000000e+00            0.000000e+00  
25%          1.000000e+00        1.000000e+01            4.000000e+00  
50%          3.000000e+00        1.300000e+01            7.000000e+00  
75%          5.000000e+00        1.600000e+01            1.500000e+01  
max          6.000000e+00        2.300000e+01            3.000000e+01

The record count for 'days_since_prior_order' is lower than for the other columns, indicating there are some nan values.

In [33]:
# Step 3
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print(col)

In [35]:
# Step 5
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_order_is_made             0
days_since_prior_order    206209
dtype: int64

There are 206209 null records in the 'days_since_prior_order' column. I belive this responds to information from the Step 2 table. The maximum value for this column is 30, this might indicate that if a client has gone longer than a month without placing an order, there is no way to reflect this on the dataframe and therefore it becomes a null value.

In [38]:
# Step 6
df_ords['days_since_prior_order'].fillna(7, inplace = True)

Because I cannot corroborate my theory (Step 5), in ortdr to continue with the exercise I imputed the median value. If I could contact the team that built the dataframe or if I had enough Python knowledge to code the question myself, I would try to discover if that is the reason why and if there is a way we can include that information. If my theory was not correct then I would go on with the imputation.

In [40]:
# Step 7
df_ords_dups = df_ords[df_ords.duplicated()]

In [41]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, hour_order_is_made, days_since_prior_order]
Index: []

There are no duplicate records in this dataframe

In [43]:
# Step 9
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_clean.csv'))